helios-client/src/main/java/com/spotify/helios/client/tls/SshAgentTlsClient.java

```diff
@@ -0,0 +1,131 @@
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.Lists;
+
+import com.spotify.sshagentproxy.Identity;
+
+import org.bouncycastle.cert.X509CertificateHolder;
+import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter;
+import org.bouncycastle.crypto.tls.Certificate;
+import org.bouncycastle.crypto.tls.CertificateRequest;
+import org.bouncycastle.crypto.tls.DefaultTlsClient;
+import org.bouncycastle.crypto.tls.HashAlgorithm;
+import org.bouncycastle.crypto.tls.SignatureAlgorithm;
+import org.bouncycastle.crypto.tls.SignatureAndHashAlgorithm;
+import org.bouncycastle.crypto.tls.TlsAuthentication;
+import org.bouncycastle.crypto.tls.TlsCredentials;
+import org.bouncycastle.crypto.tls.TlsSignerCredentials;
+
+import java.io.IOException;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.List;
+
+/**
+ * A {@link org.bouncycastle.crypto.tls.TlsClient} that uses an SSH agent identity for client
+ * authentication. An X509 certificate is generated based on the SSH agent identity, and the
+ * handshake record is sent to the local SSH agent to be signed.
+ */
+class SshAgentTlsClient extends DefaultTlsClient {
+
+  private static final SignatureAndHashAlgorithm ALGO = new SignatureAndHashAlgorithm(
+      HashAlgorithm.sha1, SignatureAlgorithm.rsa);
+
+  private final String username;
+  private final Identity identity;
+  private final RecordingTlsClientProtocol protocol;
+
+  private X509Certificate[] serverCertificates;
+
+  public SshAgentTlsClient(final String username, final Identity identity,
+                           final RecordingTlsClientProtocol protocol) {
+    super();
+    this.username = username;
+    this.identity = identity;
+    this.protocol = protocol;
+  }
+
+  public RecordingTlsClientProtocol getProtocol() {
+    return protocol;
+  }
+
+  public X509Certificate[] getServerCertificates() {
+    return serverCertificates;
+  }
+
+  @Override
+  public TlsAuthentication getAuthentication() throws IOException {
+    return new SshAgentTlsAuthentication();
+  }
+
+  private class SshAgentTlsAuthentication implements TlsAuthentication {
+
+    private final Certificate certificate;
+
+    public SshAgentTlsAuthentication() {
+      certificate = X509CertificateFactory.get(username, identity);
+    }
+
+    @Override
+    public void notifyServerCertificate(final Certificate serverCertificate)
+        throws IOException {
+      final JcaX509CertificateConverter converter =
+          new JcaX509CertificateConverter().setProvider("BC");
+
+      final List<X509Certificate> x509Certs = Lists.newArrayList();
+      try {
+        for (final org.bouncycastle.asn1.x509.Certificate cert :
+            serverCertificate.getCertificateList()) {
+          x509Certs.add(converter.getCertificate(new X509CertificateHolder(cert)));
+        }
+      } catch (CertificateException e) {
+        throw Throwables.propagate(e);
```

Comment:
since the interface has `throws IOException` should probably package any exception as that type